Inconsistent enforcement of duplicate IP addresses

### Environment
* Python version: 3.6.9
* NetBox version: 2.10.4

### Steps to Reproduce

This is with `ENFORCE_GLOBAL_UNIQUE = True` in `configuration.py`

1. Add two interfaces to a device, say "test1" and "test2" (these can be virtual)
2. Add IP address 1.2.3.4/24 to "test1" with no role
3. Add IP address 1.2.3.4/24 to "test2" with role VIP
4. Change the test1 1.2.3.4/24 IP address to role VIP
5. Change the test1 1.2.3.4/24 IP address back to no role

### Expected Behavior
I expected duplicate IP address enforcement to be consistent: either require both addresses to be VIP, or allow one normal and one VIP to share the same IP address.  (Probably the former).

In other words: either step 3 should have failed with a duplicate address error, or step 5 should have been accepted since it puts things back to how they were at step 3.

### Observed Behavior
Step 3 is accepted, but step 5 fails with:
![image](https://user-images.githubusercontent.com/44789/106595854-150c9c80-654c-11eb-91cb-01eb5037df24.png)

Is there any way to avoid it creating and deleting the entire database on every test?  On a quad-core i7 it takes over a minute just to run a single test file.

```
(netbox) root@netbox-dev:~/netbox/netbox# time python3 ./manage.py test ipam.tests.test_models
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
...F..................
======================================================================
FAIL: test_duplicate_nonunique_nonrole_role (ipam.tests.test_models.TestIPAddress)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/.venv/netbox/lib/python3.6/site-packages/django/test/utils.py", line 381, in inner
    return func(*args, **kwargs)
  File "/root/netbox/netbox/ipam/tests/test_models.py", line 266, in test_duplicate_nonunique_nonrole_role
    self.assertRaises(ValidationError, duplicate_ip.clean)
AssertionError: ValidationError not raised by clean

----------------------------------------------------------------------
Ran 22 tests in 0.177s

FAILED (failures=1)
Destroying test database for alias 'default'...

real	1m12.375s
user	0m57.308s
sys	0m0.280s
```

--- a/netbox/ipam/models.py
+++ b/netbox/ipam/models.py
@@ -734,13 +734,12 @@ class IPAddress(ChangeLoggedModel, CustomFieldModel):
                 })
 
             # Enforce unique IP space (if applicable)
-            if self.role not in IPADDRESS_ROLES_NONUNIQUE and ((
-                self.vrf is None and settings.ENFORCE_GLOBAL_UNIQUE
-            ) or (
-                self.vrf and self.vrf.enforce_unique
-            )):
+            if (self.vrf is None and settings.ENFORCE_GLOBAL_UNIQUE) or (self.vrf and self.vrf.enforce_unique):
                 duplicate_ips = self.get_duplicates()
-                if duplicate_ips:
+                if duplicate_ips and (
+                        self.role not in IPADDRESS_ROLES_NONUNIQUE or
+                        any(dip.role not in IPADDRESS_ROLES_NONUNIQUE for dip in duplicate_ips)
+                ):
                     raise ValidationError({
                         'address': "Duplicate IP address found in {}: {}".format(
                             "VRF {}".format(self.vrf) if self.vrf else "global table",

--- a/netbox/ipam/tests/test_models.py
+++ b/netbox/ipam/tests/test_models.py
@@ -260,6 +260,18 @@ class TestIPAddress(TestCase):
         self.assertRaises(ValidationError, duplicate_ip.clean)
 
     @override_settings(ENFORCE_GLOBAL_UNIQUE=True)
+    def test_duplicate_nonunique_nonrole_role(self):
+        IPAddress.objects.create(address=netaddr.IPNetwork('192.0.2.1/24'))
+        duplicate_ip = IPAddress(address=netaddr.IPNetwork('192.0.2.1/24'), role=IPAddressRoleChoices.ROLE_VIP)
+        self.assertRaises(ValidationError, duplicate_ip.clean)
+
+    @override_settings(ENFORCE_GLOBAL_UNIQUE=True)
+    def test_duplicate_nonunique_role_nonrole(self):
+        IPAddress.objects.create(address=netaddr.IPNetwork('192.0.2.1/24'), role=IPAddressRoleChoices.ROLE_VIP)
+        duplicate_ip = IPAddress(address=netaddr.IPNetwork('192.0.2.1/24'))
+        self.assertRaises(ValidationError, duplicate_ip.clean)
+
+    @override_settings(ENFORCE_GLOBAL_UNIQUE=True)
     def test_duplicate_nonunique_role(self):
         IPAddress.objects.create(address=netaddr.IPNetwork('192.0.2.1/24'), role=IPAddressRoleChoices.ROLE_VIP)
         IPAddress.objects.create(address=netaddr.IPNetwork('192.0.2.1/24'), role=IPAddressRoleChoices.ROLE_VIP)

